src/content/en/2019/css.md

```diff
@@ -31,67 +31,76 @@ Color is an integral part of theming and styling on the web. Let's take a look a
 
 Hex is the most popular way to describe color by far, with 93% usage, followed by RGB, and then HSL. Interestingly, developers are taking full advantage of the alpha-transparency argument when it comes to these color types: HSLA and RGBA are far more popular than HSL and RGB, with almost double the usage! Even though the alpha-transparency was added later to the web spec, HSLA and RGBA are supported [as far back as IE9](https://caniuse.com/#feat=css3-colors), so you can go ahead and use them, too!
 
-<figure>
-  <a href="/static/images/2019/css/fig1.png">
-    <img src="/static/images/2019/css/fig1.png" alt="Figure 1. Popularity of color formats." aria-labelledby="fig1-caption" aria-describedby="fig1-description" width="600" height="371" data-width="600" data-height="371" data-seamless data-frameborder="0" data-scrolling="no" data-iframe="https://docs.google.com/spreadsheets/d/e/2PACX-1vQO5CabwLwQ5Lj1_9bbEFnFM1qEqCorymaBHrcaNiMSJ7sYDKHUI5iish5VAS-SxN447UTW-1-5-OjE/pubchart?oid=1946838030&amp;format=interactive">
-  </a>
-  <div id="fig1-description" class="visually-hidden">Bar chart showing the adoption of HSL, HSLA, RGB, RGBA, and hex color formats. Hex is used on 93% of desktop pages, RGBA on 83%, RGB on 22%, HSLA 19%, and HSL 1%. Desktop and mobile adoption is similar for all color formats except HSL, for which mobile adoption is 9% (9 times higher).</div>
-  <figcaption id="fig1-caption">Figure 1. Popularity of color formats.</figcaption>
-</figure>
+{{ figure_markup(
+  id=1,
```

Comment:
Figure numbering is a huge part of #1138 and I see in the PR description that it's purposely not included. Do you plan to iterate on it? If not should we continue the discussion of pros/cons or technical implementation in the issue?

Reply:
Potentially. Either as part of this PR (which might make sense to avoid going at all the chapters again shortly after merging this). However I wanted feedback on what's been done so far and am still mulling over the figures in my head.

While I agree maintaining figure numbers automatically would be ideal, I think there's still huge value in what's in this PR so far even without that. The syntax is hugely simplified and now very easy to make changes consistently (e.g. change "Figure 2" to "Figure 2.2").

I'll be honest though and admit I'm still not 100% convinced on the automatic numbering. I think it would be possible to generate the numbers for the `<figure>`'s with "id=XXX" place holders and replacing with regex (much like we do now), or perhaps petty through some DOM queries. However I am less convinced about in the text, nor moving the anchors to names.

I still think naming things is hard. Yes we ideally want descriptive name for image files too which should give us those unique names but I still think writing `In Figure {{ fig_id("fig_dist_websites_performance") }} above, you can see how FCP...` is much more painful than writing `In Figure 1 above` - even if that means dealing with the consequences that some figures might need to be renumbered as we add the Figures just before launch (and only very rarely afterwards).

And, whether we like it or not, I think people may well refer to them externally ("In the Web Almanac in Figure 1.2 they showed" - though admittedly can't find an example of this!), or (by taking screenshots](https://www.stefanjudis.com/blog/highlights-of-the-http-archive-web-almanac/) so we should try to be consistent in figure numbering after launch as much as possible. I remember in #592 noticing the numbering change but that's much less likely no we don't see the templates - especially now we've taken that further with #1230 .

And, this one is probably an edge case, but I do flick between languages (e.g. /en/2019/cdn#fig-1 and /ja/2019/cdn#fig-1) so having consistency is helpful to me, rather than language specific anchors. I think the short anchor names are handy.

While on the subject, I also think the language thing is still a complexity to be solved. They can either use English (not very i18n friendly - though take your point that chapter names are not internationalised) or use their own tags - where we have to ensure consistency between references in text and figures, as well as include the English anyway for the image link.

So, in summary, I've still concerns to work through and automatic figure numbering may or may not be added. We should continue the discussion (either here or in #1138 ) while I play with how to do it technically. BUT, either way, I think there's huge value in this PR already as is.

Reply:
We could compromise and auto generate the figure numbers (like we do now) but leave the figure numbers hardcoded in the text, and still use the `figX`-style anchors?

That would leave us in same place as now, but with the figure markup drastically simplified (which is see as the main aim of this PR).

Of course that means there’s still a risk that the text references and figures get out of sync, but that’s a present risk anyway and an acceptable risk given how infrequently I expect these to change, and makes the writing simpler.

WDYT?

Reply:
OK I've had another stab at this and brought it closer to what you want:

Figure Ids are now auto-generated. They were already, so removing this actually was a reversion in that respect. I was a little uncomfortable with this anyway but think I've mitigated against breakage concerns with the addition of a a new (optional) `figure_lock` value in the chapter YAML metadata:

https://github.com/HTTPArchive/almanac.httparchive.org/blob/bead9c5941a35b8872c207dfe1299d049e05f5ba/src/content/en/2019/css.md#L11-L14

https://github.com/HTTPArchive/almanac.httparchive.org/blob/bead9c5941a35b8872c207dfe1299d049e05f5ba/src/content/en/2019/javascript.md#L11-L14

If this config is set, then `npm run generate` errors when the site is built if the figure count does not match that. So we SHOULD set this just before publishing the chapter and "lock" in the figure count:

https://github.com/HTTPArchive/almanac.httparchive.org/blob/bead9c5941a35b8872c207dfe1299d049e05f5ba/src/tools/generate/generate_figure_ids.js#L73-L82

I think this is a good safety check and hopefully alleviates both my concerns on figure ids changing, and therefore (hopefully) your concerns on using simple figure ids as anchor links.

Given that, I have kept the `fig-12` anchors. I like 'em! - short and sweet and no concerns about dupes or character encoding, or having to add them to `big-number` and `table` figures. As I say I think once we start labelling figures we should try our best to avoid changing the numbering and if we do control that, then link breakage is less of a concern.

And finally I've added the ability to reference figures in text:

https://github.com/HTTPArchive/almanac.httparchive.org/blob/bead9c5941a35b8872c207dfe1299d049e05f5ba/src/content/en/2019/css.md#L89-L99

As you can see the `ref="figref1"` is an optional argument added to the `figure_markup` call, and then references to that in the writing are automatically replaced with the appropriate figure id. In this case the text becomes:

> Unsurprisingly, In Figure 2.5 above...

Let me know your thoughts!

Reply:
The outcomes I'm hoping to achieve in this PR are abstracting away figure markup/a11y best practices and ergonomic improvements to the authoring experience. I think we hit the first goal with `figure_markup`. But on the ergonomics side I'm afraid that these changes have expanded in complexity to the point that it's actually more cumbersome to use than the current implementation. WDYT about reverting the latest `figure_lock` and `figref` changes and we can merge the `figure_markup` improvements? We can iterate on the ergonomic design asynchronously in #1138.

Reply:
I do think the `figref` is more bother than its worth and happy to remove that. Was never particularly happy with that anyway as you may have noticed 😉

However I do like the `figure_lock` stuff. The more we abstract away the figure numbering (which we have done almost completely in this) the more easy it is to accidentally change figure numbering by removing a figure and then that breaks all references in chapters and anchor links - like we almost did in #592. This is actually more likely now we no longer can see this as we stopped including the generated templates in the main PRs ages ago - and as of more recently don’t even store them in Git at all! Would you be up for keeping those? We only need to “lock” the chapters once published so authors don’t need to bother about this. WDYT?

Reply:
I don't see the benefit of `figure_lock` TBH. Contrary to my goal in #1138 to make figure management more automated/abstract, it creates another thing that needs to be manually kept in sync.

Reply:
Fair enough - it is gone!